File: observability-kit-demo-hilla/frontend/views/image-list/image-list-view.ts
import type { SelectValueChangedEvent } from '@vaadin/select/src/vaadin-select.js';
import '@vaadin/select';
import '@vaadin/horizontal-layout';
import '@vaadin/vertical-layout';
import './image-list-card.js';
import { html, type TemplateResult } from 'lit';
import { customElement } from 'lit/decorators.js';
import { repeat } from 'lit/directives/repeat.js';
import styles from './image-list-view.css';
import { Layout } from 'Frontend/views/view.js';

const $sorting = Symbol();
const $selected = Symbol();
const $onChange = Symbol();

const imageParams = {
  auto: 'format',
  fit: 'crop',
  ixid: 'MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D',
  ixlib: 'rb-1.2.1',
  q: '80',
  w: '750',
};

const images = [
  { id: '1519681393784-d120267933ba', label: 'Snow mountains under stars' },
  { id: '1512273222628-4daea6e55abb', label: 'Snow covered mountain' },
  { id: '1536048810607-3dc7f86981cb', label: 'River between mountains' },
  { id: '1515705576963-95cad62945b6', label: 'Milky way on mountains' },
  { id: '1513147122760-ad1d5bf68cdb', label: 'Mountain with fog' },
  { id: '1562832135-14a35d25edef', label: 'Mountain at night' },
].map(({ id, label }) => {
  const url = new URL(`photo-${id}`, 'https://images.unsplash.com/');

  for (const [name, value] of Object.entries(imageParams)) {
    url.searchParams.set(name, value);
  }

  return { id, label, url };
});

@customElement('image-list-view')
export default class ImageListView extends Layout {
  static override readonly styles = styles;

  private [$selected]: string;
  private readonly [$sorting] = ['Popularity', 'Newest first', 'Oldest first'];

  constructor() {
    super();
    this[$selected] = this[$sorting][0];
  }

  override connectedCallback(): void {
    super.connectedCallback();
    this.classList.add('max-w-screen-lg', 'mx-auto', 'pb-l', 'px-l');
  }

  protected override render(): TemplateResult {
    return html`<vaadin-horizontal-layout class="p-l items-center justify-between">
        <vaadin-vertical-layout>
          <h2 class="mb-0 mt-xl text-3xl">Beautiful photos</h2>
          <p class="mb-xl mt-0 text-secondary">Royalty free photos and pictures, courtesy of Unsplash</p>
          <vaadin-select label="Sort by" .items=${this[$sorting]} @value-changed=${this[$onChange]}></vaadin-select>
        </vaadin-vertical-layout>
      </vaadin-horizontal-layout>
      <ol class="p-l gap-m grid list-none m-0 p-0">
        ${repeat(
          images,
          ({ id }) => id,
          ({ id, label, url }) => html`<image-list-card src=${url} title=${label}></image-list-card>`,
        )}
      </ol>`;
  }

  private [$onChange]({ detail: { value } }: SelectValueChangedEvent) {
    this[$selected] = value;
  }
}

File: packages/java/observability-kit-demo/src/main/java/com/example/application/views/imagelist/ImageListView.java
package com.example.application.views.imagelist;

import java.util.Timer;
import java.util.TimerTask;

import com.example.application.views.MainLayout;
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.Tracer;

import com.vaadin.flow.component.HasComponents;
import com.vaadin.flow.component.HasStyle;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.component.html.H2;
import com.vaadin.flow.component.html.Main;
import com.vaadin.flow.component.html.OrderedList;
import com.vaadin.flow.component.html.Paragraph;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.progressbar.ProgressBar;
import com.vaadin.flow.component.select.Select;
import com.vaadin.flow.router.PageTitle;
import com.vaadin.flow.router.Route;
import com.vaadin.flow.server.auth.AnonymousAllowed;

@PageTitle("Image List")
@Route(value = "image-list", layout = MainLayout.class)
@AnonymousAllowed
public class ImageListView extends Main implements HasComponents, HasStyle {

    private OrderedList imageContainer;

    ProgressBar progress;

    public ImageListView() {
        progress = new ProgressBar();
        progress.setIndeterminate(true);
        add(progress);
        final UI ui = UI.getCurrent();

        Tracer trace = GlobalOpenTelemetry.getTracer("app-instrumentation",
                "1.0");
        final Span span = trace.spanBuilder("My Long Loading Task").startSpan();

        TimerTask task = new TimerTask() {
            public void run() {
                ui.access(() -> {
                    removeAll();
                    init();
                    span.end();
                });
            }
        };
        Timer timer = new Timer("Timer");

        long delay = 3000L;
        timer.schedule(task, delay);
    }

    private void init() {
        constructUI();

        imageContainer.add(new ImageListViewCard("Snow mountains under stars",
                "https://images.unsplash.com/photo-1519681393784-d120267933ba?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=750&q=80"));
        imageContainer.add(new ImageListViewCard("Snow covered mountain",
                "https://images.unsplash.com/photo-1512273222628-4daea6e55abb?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=750&q=80"));
        imageContainer.add(new ImageListViewCard("River between mountains",
                "https://images.unsplash.com/photo-1536048810607-3dc7f86981cb?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=375&q=80"));
        imageContainer.add(new ImageListViewCard("Milky way on mountains",
                "https://images.unsplash.com/photo-1515705576963-95cad62945b6?ixlib=rb-1.2.1&ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&auto=format&fit=crop&w=750&q=80"));
        imageContainer.add(new ImageListViewCard("Mountain with fog",
                "https://images.unsplash.com/photo-1513147122760-ad1d5bf68cdb?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=1000&q=80"));
        imageContainer.add(new ImageListViewCard("Mountain at night",
                "https://images.unsplash.com/photo-1562832135-14a35d25edef?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=815&q=80"));

    }

    private void constructUI() {
        addClassNames("image-list-view", "max-w-screen-lg", "mx-auto", "pb-l",
                "px-l");

        HorizontalLayout container = new HorizontalLayout();
        container.addClassNames("items-center", "justify-between");

        VerticalLayout headerContainer = new VerticalLayout();
        H2 header = new H2("Beautiful photos");
        header.addClassNames("mb-0", "mt-xl", "text-3xl");
        Paragraph description = new Paragraph(
                "Royalty free photos and pictures, courtesy of Unsplash");
        description.addClassNames("mb-xl", "mt-0", "text-secondary");
        headerContainer.add(header, description);

        Select<String> sortBy = new Select<>();
        sortBy.setLabel("Sort by");
        sortBy.setItems("Popularity", "Newest first", "Oldest first");
        sortBy.setValue("Popularity");

        imageContainer = new OrderedList();
        imageContainer.addClassNames("gap-m", "grid", "list-none", "m-0",
                "p-0");

        container.add(header, sortBy);
        add(container, imageContainer);

    }
}
